pkg/yurttunnel/pki/certmanager/certmanager.go

```diff
@@ -17,21 +17,345 @@ limitations under the License.
+	if err != nil {
+		return dnsNames, ips, err
+	}
+
+	// extract dns and ip from the service info
+	dnsNames = append(dnsNames, fmt.Sprintf("%s.%s", s.Name, s.Namespace))
+	if s.Spec.ClusterIP != "None" {
+		ips = append(ips, net.ParseIP(s.Spec.ClusterIP))
+	}
+	ips = append(ips, net.ParseIP("127.0.0.1"))
+
+	switch s.Spec.Type {
+	case corev1.ServiceTypeLoadBalancer:
+		return getLoadBalancerDNSandIP(s, dnsNames, ips)
+	case corev1.ServiceTypeClusterIP:
+		return getClusterIPDNSandIP(s, dnsNames, ips)
+	case corev1.ServiceTypeNodePort:
+		return getNodePortDNSandIP(clientset, dnsNames, ips)
+	default:
+		return dnsNames, ips,
+			fmt.Errorf("unsupported service type: %s", string(s.Spec.Type))
+	}
+}
+
+// getLoadBalancerDNSandIP gets the DNS names and IPs from the
+// LoadBalancer service
+func getLoadBalancerDNSandIP(
+	svc *corev1.Service,
+	dnsNames []string,
+	ips []net.IP) ([]string, []net.IP, error) {
+	if len(svc.Status.LoadBalancer.Ingress) == 0 {
+		return dnsNames, ips, errors.New("load balancer is not ready")
+	}
+	for _, ingress := range svc.Status.LoadBalancer.Ingress {
+		if ingress.IP != "" {
+			ips = append(ips, net.ParseIP(ingress.IP))
+		}
+
+		if ingress.Hostname != "" {
+			dnsNames = append(dnsNames, ingress.Hostname)
+		}
+	}
+	return dnsNames, ips, nil
+}
+
+// getClusterIPDNSandIP gets the DNS names and IPs from the ClusterIP service
+func getClusterIPDNSandIP(
+	svc *corev1.Service,
+	dnsNames []string,
+	ips []net.IP) ([]string, []net.IP, error) {
+	if addr, ok := svc.Annotations[constants.YurttunnelServerExternalAddrKey]; ok {
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return dnsNames, ips, err
+		}
+		ips = append(ips, net.ParseIP(host))
+	}
+	return dnsNames, ips, nil
+}
+
+// getClusterIPDNSandIP gets the DNS names and IPs from the NodePort service
+func getNodePortDNSandIP(
+	clientset kubernetes.Interface,
+	dnsNames []string,
+	ips []net.IP) ([]string, []net.IP, error) {
+	labelSelector := fmt.Sprintf("%s=false", constants.YurtEdgeNodeLabel)
+	// yurttunnel-server will be deployed on one of the cloud nodes
+	nodeLst, err := clientset.CoreV1().Nodes().List(
+		metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return dnsNames, ips, err
+	}
+	if len(nodeLst.Items) == 0 {
+		return dnsNames, ips, errors.New("there is no cloud node")
+	}
+	for i, addr := range nodeLst.Items[0].Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			ips = append(ips, net.ParseIP(addr.Address))
+		}
+		// there is no qualified address (i.e. NodeInternalIP)
+		if i == len(nodeLst.Items[0].Status.Addresses)-1 {
+			return dnsNames, ips, errors.New("can't find node IP")
+		}
+	}
+	return dnsNames, ips, nil
+}
+
+// NewYurttunnelAgentCertManager creates a certificate manager for
+// the yurttunel-agent
+func NewYurttunnelAgentCertManager(
+	clientset kubernetes.Interface) (certificate.Manager, error) {
+	nodeIP, err := getAgentNodeIP(clientset, os.Getenv("NODE_NAME"))
```

Comment:
tunnel agent will run on edge node with HostNetwork mode, so pod ip of tunnel agent is the same of node ip, so would you use pod ip to refactor getAgentNodeIP?

Reply:
In order to use the pod IP, we should expose the podIP to the container by adding following env setting in the yurttunel-agent.yaml?